File: apps/api/src/application/products/create-product.ts
import { normalizeCategoryName, normalizeSku } from '@/domain/product/product'

import type { CategoryRepository } from '@/application/ports/category-repository'
import type { ProductRepository } from '@/application/ports/product-repository'
import type { Product } from '@/domain/product/product'
import type { CreateProduct as CreateProductInput } from '@ecommerce/shared'

export class CreateProduct {
  constructor(
    private readonly products: ProductRepository,
    private readonly categories: CategoryRepository,
  ) {}

  async execute(input: CreateProductInput): Promise<Product> {
    const category = input.category
      ? await this.categories.findOrCreate(normalizeCategoryName(input.category))
      : null
    return this.products.create({
      sku: normalizeSku(input.sku),
      name: input.name,
      description: input.description ?? null,
      price: input.price,
      stock: input.stock,
      weightKg: input.weightKg ?? null,
      categoryId: category?.id ?? null,
    })
  }
}

File: apps/api/src/ecommerce_api/application/products/create_product.py
from ecommerce_api.application.ports.category_repository import CategoryRepository
from ecommerce_api.application.ports.product_repository import NewProduct, ProductRepository
from ecommerce_api.application.schemas.product import CreateProductInput
from ecommerce_api.domain.product import Product, normalize_category_name, normalize_sku


class CreateProduct:
    def __init__(self, products: ProductRepository, categories: CategoryRepository) -> None:
        self._products = products
        self._categories = categories

    async def execute(self, data: CreateProductInput) -> Product:
        category = (
            await self._categories.find_or_create(normalize_category_name(data.category))
            if data.category
            else None
        )
        return await self._products.create(
            NewProduct(
                sku=normalize_sku(data.sku),
                name=data.name,
                description=data.description,
                price=data.price,
                stock=data.stock,
                weight_kg=data.weight_kg,
                category_id=category.id if category else None,
            )
        )
